### PR TITLE
gh-93283: Improve error message for f-string with invalid conversion character

### LINKED ISSUE #93283: Unexpected whitespace handling in f-string !r
**Bug report**

The handling of whitespace around the argument in literal string interpolation (f-string) is unexpected when a conversion specifier (e.g. !r) is applied.

Consider
```py
f"{None}"  # OK
f"{ None }"  # OK
f"{None!r}"  # OK
f"{ None!r }"  # SyntaxError

# details
f"{ None!r}  # OK
f"{None !r}  # OK
f"{None!r }  # SyntaxError
```

PEP-498 states:

> [Leading and trailing whitespace in expressions is ignored](https://peps.python.org/pep-0498/#leading-and-trailing-whitespace-in-expressions-is-ignored)
For ease of readability, leading and trailing whitespace in expressions is ignored. This is a by-product of enclosing the expression in parentheses before evaluation.

I suppose that the text above may be technically accurate, but is misleading to common folk:
In the example above, `None` is enclosed in parentheses `(None)` but not `None!r`.

There are no mentions of whitespace in the rest of the PEP.

Formal reproducer:

```command
> python3.11 -c 'print(f"{ None!r }")'
  File "<string>", line 1
    print(f"{ None!r }")
                       ^
SyntaxError: f-string: expecting '}'
```

**Your environment**

- CPython versions tested on: 3.8, 3.9, 3.10, 3.11a2, 3.11a7
- Operating system and architecture: Linux, MacOS

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
#93283.

### CODE REVIEW

**Review by serhiy-storchaka — COMMENTED** [review]
What error should be for `f'{3!g'`? "invalid conversion character" or "expecting '}'"?

What error should be for `f'{3!:'`? "missed conversion character" or "expecting '}'"?

**Comment by AA-Turner** [comment]
> 

Personally, I'd expect `expecting '}'`, and then after I add the closing `}`, I'd then expect `invalid conversion character 'g'`.

A

**Comment by pablogsal** [comment]
`expecting '{'` is much better IMHO.

**Comment by serhiy-storchaka** [comment]
It is easy for `f'{3!g'`, but for `f'{3!:'` the conversion character is raised before parsing of the format string is even started.

File: Lib/test/test_fstring.py [hunk 1]
```diff
@@ -1012,19 +1014,28 @@ def test_conversions(self):
...
-        self.assertAllRaise(SyntaxError, "f-string: expecting '}'",
-                            ["f'{x!s{y}}'",
-                             "f'{3!ss}'",
+        for conv in 'g', 'A', '3', 'G', '!', ' s', 's ', 'ж':
+            self.assertAllRaise(SyntaxError,
+                                "f-string: invalid conversion character %r: "
+                                "expected 's', 'r', or 'a'" % conv,
+                                ["f'{3!" + conv + "}'"])
+
+        self.assertAllRaise(SyntaxError,
```
Comment (ericvsmith) [inline Lib/test/test_fstring.py:1035]: Somewhere we should probably add whitespace both before and after the conversion character: "s ", " s", " s ". It looks like that got dropped.

**Comment by serhiy-storchaka** [comment]
> Btw we are just working on this in the "porting f-strings to the grammar" project in case you have time to help @serhiy-storchaka

The idea is good at first glance, I thought about it, but it looks too complicated. The main problem is that the closing quote must match the opening one. It can be solved, but it needs significant effort. There are easier and more important problems.

**Review by ericvsmith — APPROVED** [review]
LGTM.

**Pull request merged**